src/mrinufft/trajectories/tools.py

```diff
@@ -369,6 +380,290 @@ def unepify(trajectory: NDArray, Ns_readouts: int, Ns_transitions: int) -> NDArr
+
+    assert (
+        kspace_start_loc.shape
+        == kspace_end_loc.shape
+        == start_gradients.shape
+        == end_gradients.shape
+    ), "All input arrays must have shape (nb_shots, nb_dimension)"
+    if nb_raster_points is None:
+        # Calculate the number of time steps based on the area needed
+        n_ramp_down, n_ramp_up, n_plateau, gi = get_gradient_times_to_travel(
+            kspace_end_loc=kspace_end_loc,
+            kspace_start_loc=kspace_start_loc,
+            end_gradients=end_gradients,
+            start_gradients=start_gradients,
+            gamma=gamma,
+            raster_time=raster_time,
+            gmax=gmax,
+            smax=smax,
+        )
+        # Extra 2 buffer samples
+        nb_raster_points = np.max(n_ramp_down + n_ramp_up + n_plateau) + 2
+
+    area_needed = (kspace_end_loc - kspace_start_loc) / gamma / raster_time
+
+    def solve_gi_fixed_N(gs, ge, area):
+        def _residual(gi):
+            n_down, n_up = _calculate_ramps(gs, ge, gi, smax, raster_time, buffer=1)
+            n_pl = nb_raster_points - n_down - n_up
+            if n_pl < 0:
+                return np.abs(n_pl)  # Penalize this
+            area_expr = _calculate_area(gs, ge, gi, n_down, n_up, n_pl)
+            return np.abs(area - area_expr)
+
+        res = minimize_scalar(
+            _residual, bounds=(-gmax, gmax), method="bounded", options={"xatol": 1e-10}
+        )
+        if not res.success:
+            raise RuntimeError(f"Minimization failed: {res.message}")
+        return res.x
+
+    gi = Parallel(n_jobs=n_jobs)(
+        delayed(solve_gi_fixed_N)(
+            start_gradients[i, j],
+            end_gradients[i, j],
+            area_needed[i, j],
+        )
+        for i in range(start_gradients.shape[0])
+        for j in range(start_gradients.shape[1])
+    )
+    gi = np.reshape(gi, start_gradients.shape)
+    n_ramp_down, n_ramp_up = _calculate_ramps(
+        start_gradients,
+        end_gradients,
+        gi,
+        smax,
+        raster_time,
+        ceil=True,
+    )
+    n_plateau = nb_raster_points - n_ramp_down - n_ramp_up
+    gi = _calculate_gi(
+        start_gradients, end_gradients, n_ramp_down, n_ramp_up, n_plateau, area_needed
+    )
+    nb_shots, nb_dimension = kspace_end_loc.shape
+    G = np.zeros((nb_shots, nb_raster_points, nb_dimension), dtype=np.float32)
+    for i in range(nb_shots):
+        for d in range(nb_dimension):
+            start = 0
+            G[i, : n_ramp_down[i, d], d] = np.linspace(
+                start_gradients[i, d], gi[i, d], n_ramp_down[i, d], endpoint=False
+            )
+            start += n_ramp_down[i, d]
+            if n_plateau[i, d] > 0:
+                G[i, start : start + n_plateau[i, d], d] = gi[i, d]
+                start += n_plateau[i, d]
+            G[i, start : start + n_ramp_up[i, d], d] = np.linspace(
+                gi[i, d], end_gradients[i, d], n_ramp_up[i, d], endpoint=False
+            )
+    return G
```

Comment:
linspace is vector-friendly:

```suggestion
    G = np.zeros((nb_shots, nb_raster_points, nb_dimension), dtype=np.float32)
    for i in range(nb_shots):
        start = n_ramp_down[i,0]
        G[i, : start] = np.linspace(
            start_gradients[i], gi[i], n_ramp_down[i], endpoint=False, axis=-1
        )
        if n_plateau[i, d] > 0:
            G[i, start : start + n_plateau[i, 0]] = gi[i]
            start += n_plateau[i, 0]
        G[i, start : start + n_ramp_up[i, 0]] = np.linspace(
            gi[i], end_gradients[i], n_ramp_up[i, 0], axis=-1, endpoint=False
        )
    return G
```

Reply:
I dont understand this, the ramps and gi values can change for each dimension as well. Sadly there is no way to vectorize it. However, this is still pretty fast.